python/paddle/fluid/incubate/fleet/collective/__init__.py

```diff
@@ -48,8 +48,25 @@ def stop_worker(self):
             "You should not call 'stop_worker' method for collective mode.")
 
     def distributed_optimizer(self, optimizer, strategy=None):
-        self._optimizer = CollectiveOptimizer(optimizer, strategy)
-        return self._optimizer
+        if strategy == None:
+            self._optimizer = CollectiveOptimizer(optimizer, strategy)
+            return self._optimizer
+        if strategy.collecitve_op_based:
+            if strategy.use_fp16:
+                self._optmizer = \
+                    MixedPrecisionOptimizer(optmizer, strategy)
+            else:
+                self._optimizer = \
+                    FullPrecisionOptimizer(optmizer, strategy)
+        if strategy.local_sgd:
+            if strategy.use_fp16:
+                # should have an independent optimizer inherited from
+                # CollectiveOpBasedOptimizer
+                pass
+            else:
+                # should have an independent optimizer inherited from
+                # CollectiveOpBasedOptimizer
+                pass
```

Comment:
could we use decorator pattern to replace these ifelse? e.g. if we require algorithm chain as momentum + fp16 + localsgd, we declare
`opt = LocalSGD(MixedPrecision(Momentum(learning_rate=learning_rate, momentum=0.9)))`

Reply:
a factory or decorator can be design as free. Generally, a user only need to define the strategy to do distributed optimization, and instance of DistributedOptimizer is determined by strategy a user defined.